core/modules/ccontrol/testCControl.cc

```diff
@@ -615,6 +619,37 @@ BOOST_DATA_TEST_CASE(antlr_compare, QUERIES, query) {
 }
 
 
+BOOST_AUTO_TEST_CASE(not_like) {
+    // NOT LIKE can not be tested against the antlr2 parser because that parser can't parse NOT LIKE.
+    // Instead, we test NOT LIKE by generating a statement with LIKE, and then manually change the
+    // intermediate representation of that statement to be NOT LIKE, and compare that with a generated
+    // NOT LIKE statement.
+    // We know the LIKE statement is correct because we have a test that compares it with the
+    // antlr2-generated LIKE statement, in antlr_compare.
+    auto parserWithNot = parser::SelectParser::newInstance(
+            "select shortName from Filter where shortName NOT LIKE 'Z'", parser::SelectParser::ANTLR4);
+    BOOST_REQUIRE_NO_THROW(parserWithNot->setup());
+    auto selectStatementWithNot = parserWithNot->getSelectStmt();
+    BOOST_REQUIRE(selectStatementWithNot != nullptr);
+
+    auto parserWithoutNot = parser::SelectParser::newInstance(
+            "select shortName from Filter where shortName LIKE 'Z'", parser::SelectParser::ANTLR4);
+    BOOST_REQUIRE_NO_THROW(parserWithoutNot->setup());
+    auto selectStatementWithoutNot = parserWithoutNot->getSelectStmt();
+    BOOST_REQUIRE(selectStatementWithoutNot != nullptr);
+
+    // flip the 'not' on the 'without' statement to make it 'with not'
+    auto whereClauseRef = selectStatementWithoutNot->getWhereClause();
+    auto orTerm = std::dynamic_pointer_cast<query::OrTerm>(whereClauseRef.getRootTerm());
+    auto andTerm = std::dynamic_pointer_cast<query::AndTerm>(orTerm->_terms[0]);
+    auto boolFactor = std::dynamic_pointer_cast<query::BoolFactor>(andTerm->_terms[0]);
+    auto likePredicate = std::dynamic_pointer_cast<query::LikePredicate>(boolFactor->_terms[0]);
+    likePredicate->hasNot = true;
+    // now the the LIKE and the NOT LIKE selectStatements should be exactly the same:
+    BOOST_REQUIRE_EQUAL(*selectStatementWithNot, *selectStatementWithoutNot);
+}
+
+
```

Comment:
I'm a bit confused here. The comment at the beginning of this test case is mentioning ANTLR2, meanwhile I'm only seeing ANTLR4 when instantiating select statements by `parser::SelectParser::newInstance`. How ANTLR2 is supposed to be tested? Is there any internal switch in the implementation of `parser::SelectParser::newInstance` which would return an older version of the parser?

Reply:
I added more comments to the code, hopefully it's clearer: we can't compare the antlr4-generated IR to the antlr2-generated IR because antlr2 couldn't parse NOT LIKE. so we create the Like statement, and change the NOT var from 0 to 1 (to make it NOT, instead of not-NOT ;-) )